### ISSUE TITLE
Facebook ANE Crash On initial FB.logManager.requestPermission call on Android

### ISSUE THREAD

**Issue by hybridmindset** [myflashlab/facebook-ANE#18]
I am compiling my project with your commercial ANE everything compiles correctly. I am using: CommonV4.0.ane
Facebook.ane
Adobe AIR SDK v18
Installed on Android 5.1.1

My app runs just fine. I initially did this with the demo ane and when I made the FB.logManager.requestPermission call it presented the dialog saying "this is a demo ANE" now that I have swapped out the commercial ANE and recompiled when I make this same api call the app instantly crashes. Is there a different setup process or workflow for the commercial ANE? 

I call:

FB.getInstance(appID);
then 
FB.logManager.requestPermission(LogManager.WITH_READ_PERMISSIONS, Permissions.public_profile, Permissions.user_friends, Permissions.email);

And the app crashes on the FB.logManager.requestPermission call. I'm using v3.9.1 of your Facebook ANE.

**Comment by hybridmindset** [comment]
None of your tutorial links work right now http://myappsnippet.com/adobe-air-facebook-sdk-integration-part-1

**Comment by hybridmindset** [comment]
Replacing http://myappsnippet.com/ in the url with http://www.myflashlabs.com/ seems to fix the links. Now maybe reading through the tutorials I can see what I am missing.

**Comment by hybridmindset** [comment]
Does this ANE not work with stage3D (hardware accelerated) AIR apps?

**Comment by myflashlab** [comment]
Hi @hybridmindset would you please be more specific on what error message you are receiving? please note that the demo version of the ANE here on Github works exactly the same as the commercial version except for the dialog which opens at first.

the main cause of most of the crash reasons are when you don't have your manifest .xml file set correctly. Please make sure you are setting everything right by comparing to our sample here https://github.com/myflashlab/facebook-ANE/blob/master/FD/application.xml

you should be able to see all the tutorials here http://www.myflashlabs.com/product/facebook-ane-adobe-air-native-extension/

if you still have any problem initializing the extension, maybe we can help more if you are sending us your project demo so we can investigate this more nearly.

**Comment by hybridmindset** [comment]
@myflashlab there is no error message. My app compiles and runs just fine only after calling the requestPermission api of your extension the app completely stops. The phone just simply says "your app has stopped". I followed all tutorials, created the correct hash, setup the Facebook app correctly and everything. Here is my manifest file. Its just like your example. 
[TestFacebook-app_android_DEFAULT.txt](https://github.com/myflashlab/facebook-ANE/files/75179/TestFacebook-app_android_DEFAULT.txt)

**Comment by myflashlab** [comment]
@hybridmindset There are actually a couple of mistakes I quickly spotted which I mention below:

1) according to the tutorials, you shouldn't use capital words in your app package name for example you should have com.facebooktest.gbs instead of com.FacebookTest.GBS

2) on line 47 you have ```<data android:scheme="air.com.doitflash.exfacebook2" />``` this is the demo package name! you should had entered ```<data android:scheme="air.com.facebooktest.gbs" />```

These are the obvious points. then you should make sure the facebook id you have put here com.facebook.app.FacebookContentProvider174154009271869 is correct according to your facebook app settings?

you also need to make sure you have added at least the commonDependenciesV4.0.ane to your project not the lower versions.

if you checked the above points and you are sure they are all set according to the tutorials and you still have the problem, we suggest you to run the demo app here on github directly and apply your own package settings to make sure everything is working fine.

if you still have any problem, you can send us your demo project so we can verify the settings.

**Comment by hybridmindset** [comment]
@myflashlab I've searched for the word "capital" in a couple of the tutorials pages and I don't see where it mentions anything about not using capitals in the package name. This should be in huge letters in bold on the top of the manifest setup tutorial (Tut 8) or on the very first tutorial. Also you example manifest in the tutorial has "air.com.doitflash.exfacebook2" as the package name in the sample manifest. It would be a lot clearer if you just said to replace your own package name there by putting a little comment right above it in the manifest just like you have for the other areas in the manifest that need to be replaced.

Regardless, I made those changes to the app bundle id, and checked the facebook app Id, and everything you mentioned above. The app is still stopping once calling that api. So I will try to compile and run your demo app. Not sure this will make a difference because I compiled and tested my app with your demo ANE and I did get a dialog prompt with a message when calling the requestPermission api. However when I pressed the button in my app to call it a second time the app would stop. I will see if I can get my sample project files together for you to compile and run.

**Comment by myflashlab** [comment]
The reason of the crash can't be the package name anyhow but you would have faced the problem later. I will make sure the tutorials will get updated soon. Thanks for the heads up. yet, it was written here in this tutorial: http://www.myflashlabs.com/open-adobe-air-app-browser-pass-parameters/

if you can get your sample project in FlashDevelop, it will be a lot easier for us to investigate and find the problem you are facing.

a quick question, tell me your device model and Android SDK + your currently installed Facebook app verision you are workingwith.

**Comment by hybridmindset** [comment]
Galaxy S5 Model Number SM-G900T running android 5.1.1. I compiled using Adobe AIR SDK v 18.0 so whatever Android SDK is embedded in that is what is being used. I am not using a separate Android SDK to compile. The Facebook app installed on my phone is version 57.0.0.18.136

**Comment by myflashlab** [comment]
everything seems fine. Try sending us your sample project so we can duplicate the problem and help you further.

**Comment by hybridmindset** [comment]
@myflashlab Here is the link to my test project. I Don't use FlashDevelop so put together an ant script that you can run via the command line. All you need to do is replace the paths in the build.properties file located in the ant directory to point to your ant, AIR sdk, and java runtime directories. Make sure to also replaced the absolute path to the folder that you unzip the TestFacebook contents into. You can run :

$ ant package

from the command line and an apk will be generated for you and placed in the release folder. There is also a "compile" target if you want to re-compile the source. It is already compiled for you in the zip file so it is not necessary unless you make changes to the source code in the build folder.

https://s3.amazonaws.com/hmpublicfiles/TestFacebook.zip

I hope you guys can figure out what's going on otherwise I need to use another ANE.

**Comment by myflashlab** [comment]
could you also send the buggy .apk file you already have? we should do a quick test on that faster that running the demo project

**Comment by hybridmindset** [comment]
[TestFacebook.apk.zip](https://github.com/myflashlab/facebook-ANE/files/85405/TestFacebook.apk.zip)

**Comment by myflashlab** [comment]
I checked all the .as files but couldn't find where you have tried to initialized the FB ANE. 

I don't see any imports to any of the

```
	import com.myflashlab.air.extensions.facebook.access.Permissions;
	import com.myflashlab.air.extensions.facebook.access.LogManager;
	import com.myflashlab.air.extensions.facebook.FB;
	import com.myflashlab.air.extensions.facebook.FBEvent;
	import com.myflashlab.air.extensions.facebook.LikeBtn;
	import com.myflashlab.air.extensions.facebook.sharing.ShareLink;
```

am I missing something? please advice.

**Comment by hybridmindset** [comment]
That is inside of the FBPlugin.swc. I can't share all source code.  The line below is simply what is called on your Facebook api:

FB.logManager.requestPermission(LogManager.WITH_READ_PERMISSIONS, Permissions.public_profile, Permissions.user_friends, Permissions.email);

after executing:

_facebookNativeInstance = FB.getInstance(appID);
			
FB.logManager.addEventListener(FBEvent.LOGIN_DONE, onFBResponse);
FB.logManager.addEventListener(FBEvent.LOGIN_CANCELED, onFBResponse);
FB.logManager.addEventListener(FBEvent.LOGIN_ERROR, onFBResponse);

**Comment by hybridmindset** [comment]
The app stops right after calling FB.logManager.requestPermission

**Comment by myflashlab** [comment]
the error message while debugging the .apk says:

```Unable to resume activity {air.com.facebooktest.gbs/com.facebook.FacebookActivity}: The SDK has not been initialized, make sure to call FacebookSdk.sdkInitialize() first.```

which means you have not correctly initialized the extension but as we can't see the code, we can't decide.

Let us help you in another way. You said you are not using FlashDevelop, right? Tell me if you are using FlashBuilder? so we can just send you a working FlashBuilder sample?

I think this approach will be a lot easier to fix the problem.

**Comment by hybridmindset** [comment]
Yes I use flashbuilder. Where is there an initialize api in your provided code? Your documentation and example project code in this repo only has a call to FB.getInstance(appID);.

**Comment by myflashlab** [comment]
yes that's the call. 

tell me if you were able to extract the hash key successfully?

```actionscript
FB.getInstance("000000000000");
trace("hash key = ", FB.hashKey);
```

**Comment by hybridmindset** [comment]
Well that is what my code calls. It calls the FB.getInstance first. That is obviously not initializing the API in the ANE. I don't have the AS3 source code from your ANE so I can debug either. Yes I have the hash key. I even printed it out to verify that it matched my own manually generated hash. For the cert in the sample project the hash is UG7KvL4U4LhKUhqFFeYMth7ueak=

**Comment by hybridmindset** [comment]
Do I have to wait until I get the initialized event back from the ANE before calling any other API?

**Comment by hybridmindset** [comment]
Actually there is on initialized event on your FBEvent class so scratch that.

**Comment by myflashlab** [comment]
> Do I have to wait until I get the initialized event back from the ANE before calling any other API?

No, there's no such Event at all. the initialization happens instantly.

Anyhow, we are investigating on this on a FlashBuilder setup and will get back to you very shortly.

**Comment by myflashlab** [comment]
I confirm that the ANE is working just as expected. here's a FlashBuilder Sample demo project. you can run the project and build an .apk it will install on the device and runs fine exactly like the FlashDevelop demo we have. I think you should work on this demo which is in FlashBuilder version and find the possible little mistake in your setup. http://myflashlabs.com/showcase/Facebook_Demo_V3.9.1_FlashBuilder.zip (tested with release and demo ANEs both)

yet, here are some hints that you might find helpful.
1) make sure your .apk is running in captive mode.
2) when debugging in FlashBuilder, the IDE automatically changed the package name of the device, make sure it's not changed.

Let me know how you going.

**Comment by hybridmindset** [comment]
Did you use Adobe AIR sdk version 18.0 to compile?

**Comment by myflashlab** [comment]
I used 19 but i think 18 should also be ok. Have you tried with 19 or even
20?

On Monday, January 11, 2016, Lavon Woods <notifications@github.com> wrote:

> Did you use Adobe AIR sdk version 18.0 to compile?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/myflashlab/facebook-ANE/issues/18#issuecomment-170643436>
> .
>


-- 
Sent from iPhone

**Comment by hybridmindset** [comment]
No I am using AIR 18. That may be why you are not seeing the issue.

**Comment by myflashlab** [comment]
Please run the demo FlashBuilder project we sent you. If you manage to run
that, you'll be able to know what could have gone wrong in your own setup.
If you have any problems though, do feel free to ask. We'd be happy to help.

On Monday, January 11, 2016, Lavon Woods <notifications@github.com> wrote:

> No I am using AIR 18. That may be why you are not seeing the issue.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/myflashlab/facebook-ANE/issues/18#issuecomment-170652592>
> .
>


-- 
Sent from iPhone

**Comment by hybridmindset** [comment]
Running your project doesn't help me to fix my project. My project compiles and runs fine on my Android device. There isn't anything wrong with my setup. If you haven't even tried running your ANE with version 18 of the sdk how can you say it works?

**Comment by myflashlab** [comment]
Because when we were developing the ane, the latest version of air was 18!
Anyway, why don't you like to upgrade to the latest AirSDK anyway?! We
recommend you to upgrade your air sdk, that might help.

On Monday, January 11, 2016, Lavon Woods <notifications@github.com> wrote:

> Running your project doesn't help me to fix my project. My project
> compiles and runs fine on my Android device. There isn't anything wrong
> with my setup. If you haven't even tried running your ANE with version 18
> of the sdk how can you say it works?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/myflashlab/facebook-ANE/issues/18#issuecomment-170654568>
> .
>


-- 
Sent from iPhone

**Comment by hybridmindset** [comment]
Well I've tried to get it working and it doesn't work with my project. I've tried everything. Please issue a refund I will try another ANE. Thanks

**Comment by myflashlab** [comment]
sorry to hear that. well, we did all we could and everything is working fine from our side as we see and as by many other devs who are using this ANE. for refund questions you need to contact the sale team on our website http://www.myflashlabs.com/contact/ I'm closing this issue now.

**Comment by hybridmindset** [comment]
@myflashlab the problem is I can't get my project working with your ANE and I'm not getting any help from you. If you can't even get my project to work with your ANE, how will I?

**Comment by myflashlab** [comment]
@hybridmindset We kindly requested you to send us a demo project in FlashBuilder so we can easily run in our environment. you don't have to send us your whole project. For us to be able to help you, we need to be able to duplicate the problem first. You said, you are using FlashBuilder to develop your project, well, we provided you with a working demo FlashBuilder project above. your project shouldn't be different than the demo project we sent, right? If it doesn't run in your project, there must be something wrong there not the ANE. we are all programmers, right? this is a simple logic. I hope I am making sense.

Again, I repeat myself: send us a demo project (a real FlashBuilder project this time! So we can just open in a FlashBuilder IDE) and we can run your demo and be able to duplicate the problem. send us all the necessary information we need to run the demo project. we are willing to help you and I'm sure you agree that we have been replying back your message VERY fast. that's because we intent to help you as fast as possible.

You can contact us through our website and link to this page, then we'll get back to you again from there.

**Comment by hybridmindset** [comment]
Every AIR dev does not use Flashbuilder or Flashdevelop. If you are a real dev you should be able to compile, package, and deploy AIR apps without the need for Flashbuilder. It can all be done from the command line or using an ANT script. Which is what I sent you. I sent everything you needed to compile and package the APK from my source. I even stripped my project down further and it still crashes. Why can't your ANE run without an IDE? What is Flashbuilder adding to the compilation process that I don't have in my ant script that does the same thing which I sent over?

**Comment by hybridmindset** [comment]
@myflashlab I finally got your sample flashbuilder project to run in debug mode on my device and I saw the Facebook login screen. However when I export and install a captive runtime production build with the same project the screen is blank. I see a loading indicator and then just a blank screen. I don't know what is the difference between your project. 

Something else that is weird is when I have your demo project app installed on my device and I try to install my own compiled project APK to the same device it fails saying "INSTALL_FAILED_CONFLICTING_PROVIDER". That should not be the case if these two projects are using two different app ids.

**Comment by myflashlab** [comment]
> "INSTALL_FAILED_CONFLICTING_PROVIDER"

did you Google that before posting here? https://www.google.com/search?q=facebook+%22INSTALL_FAILED_CONFLICTING_PROVIDER%22

You are certainly using the same provider id in your app. which is wrong!

I really suggest you to work on a clean project and study the documentations with patience. If there are error messages you are facing, you should consider doing some Googling. you see, our job here is to build the ANE but you can't expect us to solve problems related to the native side. for example the provider conflict you mentioned. if you do a quick search, you would have seen the solution.

I must add that our job is also not to teach devs how to use ANEs in AdobeAir! there are many resources you can find online.

Our job, is to develop ANEs and make sure they work flawlessly and help developers run the ANEs on their local system.

I am going to lock down this conversation now as the discussion has gone totally out of hand on many unrelated subjects. If you have any other questions, please open a new ticket.
